lib/hash.ts

```diff
@@ -0,0 +1,57 @@
+// This module implements the "double hash" signature specification described in docs/signature_specification.md
+import * as crypto from 'crypto';
+import { FileContent, FilePath, FullFileHash, SignatureResult } from './types';
+import { isBinary } from './utils/binary';
+import { removeWhitespaces } from './utils/format';
+
+enum DigestFormat {
+  BASE64 = 'base64',
+  HEX = 'hex',
+}
+
+enum HashAlgorithm {
+  MD5 = 'md5',
+  OTHER = 'other',
+}
+
+const usedHashAlgorithm: HashAlgorithm = HashAlgorithm.MD5;
+
+export async function getHashSignature(
+  path: FilePath,
+  content: FileContent,
+): Promise<SignatureResult> {
+  const hashes: FullFileHash[] = await Promise.all([
+    computeHash(content),
+    computeUHash(content),
+  ]);
+
+  return {
+    path: path,
+    size: content.length,
+    hashes_ffm: hashes,
+  };
+}
+
+export async function computeHash(content: FileContent): Promise<FullFileHash> {
+  const hash = crypto.createHash(usedHashAlgorithm).update(content);
+  const base64Digest = hash.digest(DigestFormat.BASE64).replace(/=/g, '');
+
+  return {
+    data: base64Digest,
+    format: 1,
+  };
+}
+
+export async function computeUHash(
+  content: FileContent,
+): Promise<FullFileHash> {
+  const file = isBinary(content) ? content : removeWhitespaces(content);
+
+  const hash = crypto.createHash(usedHashAlgorithm).update(file);
+  const hexDigest = hash.digest(DigestFormat.HEX);
```

Comment:
This should be:

```
const hexDigest = hash.digest(DigestFormat.HEX).slice(0, 24);
```

Reply:
Good catch! We changed the format of the digest to 24 chars as stated in the docs:

-  strip away the UTF-8 BOM if present
- remove a set of "whitespace" bytes from the data
- generate an md5 hash over the remaining data
- **truncate the hash to 24 hex characters**